src/Sarif.Multitool/Rules/SARIF2005.ProvideHelpfulToolInformation.cs

```diff
@@ -0,0 +1,93 @@
+        };
+
+        public override FailureLevel DefaultLevel => FailureLevel.Warning;
+
+        public override string Id => RuleId.ProvideHelpfulToolInformation;
+
+        protected override IEnumerable<string> MessageResourceNames => new string[]
+        {
+            nameof(RuleResources.SARIF2005_ProvideHelpfulToolInformation_Warning_ProvideToolVersion_Text),
+            nameof(RuleResources.SARIF2005_ProvideHelpfulToolInformation_Warning_ProvideConciseToolName_Text),
+            nameof(RuleResources.SARIF2005_ProvideHelpfulToolInformation_Warning_UseNumericToolVersions_Text)
+        };
+
+        protected override void Analyze(Tool tool, string toolPointer)
+        {
+            if (tool == null)
+            {
+                return;
+            }
+
+            if (tool.Driver != null)
+            {
+                AnalyzeToolDriver(tool.Driver, toolPointer);
+            }
+        }
+
+        private void AnalyzeToolDriver(ToolComponent toolComponent, string toolPointer)
+        {
+            // ProvideConciseToolName: Ensure that tool.driver.name isn't more than 3 words long
+            if (!string.IsNullOrEmpty(toolComponent.Name))
+            {
+                const int maxWordCount = 3;
+                int wordCount = toolComponent.Name.Split(' ').Length;
+                if (wordCount > maxWordCount)
+                {
+                    string driverPointer = toolPointer.AtProperty("driver").AtProperty("name");
+                    LogResult(
+                        driverPointer,
+                        nameof(RuleResources.SARIF2005_ProvideHelpfulToolInformation_Warning_ProvideConciseToolName_Text),
+                        wordCount.ToString(),
+                        maxWordCount.ToString());
+                }
+            }
+
+            // ProvideToolVersion: Either tool.driver.version or tool.driver.semanticVersion should be there.
+            if (string.IsNullOrWhiteSpace(toolComponent.Version) && string.IsNullOrWhiteSpace(toolComponent.SemanticVersion))
+            {
+                LogResult(
+                        toolPointer,
+                        nameof(RuleResources.SARIF2005_ProvideHelpfulToolInformation_Warning_ProvideToolVersion_Text));
+            }
+            else
+            {
+                // UseNumericToolVersions
+                if (!string.IsNullOrWhiteSpace(toolComponent.Version))
+                {
+                    AnalyzeVersion(toolComponent.Version, toolPointer.AtProperty("driver").AtProperty("version"));
+                }
+                else
+                {
+                    AnalyzeVersion(toolComponent.SemanticVersion, toolPointer.AtProperty("driver").AtProperty("semanticVersion"));
+                }
+            }
+        }
+
+        private void AnalyzeVersion(string version, string pointer)
+        {
+            Regex regex = new Regex(@"^\d+\.\d+.*");
```

Comment:
>Regex regex = new Regex(@"^\d+\.\d+.*"); [](start = 12, length = 40)

We only want to construct this regex once. So, outside of this method, put:
```
private static readonly s_versionRegexe = new Regex(@"^\d+\.\d+.*", RegexOptions.Compiled | RegexOptions.CultureInvariant);
```
and use that here.

Reply:
Can i change to RegexVersion? or something like that?

---
In reply to: [444575685](https://github.com/microsoft/sarif-sdk/pull/1926#discussion_r444575685) [](ancestors = 444575685)

Reply:
1) I made a typo. It should be `s_versionRegex`. I added an extra `e` at the end, sorry.

---
In reply to: [444577200](https://github.com/microsoft/sarif-sdk/pull/1926#discussion_r444577200) [](ancestors = 444577200,444575685)